# Capstone Project - The Battle of Neighborhoods

# Brooklyn Food Truck Location Recommendations

# I. PURPOSE

This document provides the details of my peer reviewed assignment for the IBM Data Science Professional Certificate Capstone project.

# II. INTRODUCTION

New York City is the largest city in U.S.A by population, which also encompasses five divisions called boroughs.  Of the five boroughs we will focus on Brooklyn, which is known for its night life and great food.

# III. Objective

In this project, we will studies the areas of Brooklyn in order to determine which neighborhoods have the highest concentration of bars in order to solve our business problem.  

## A description of the problem and discussion of the background. (15 marks)

Our business problem is that we are working with a Food Truck business owner who is trying to determine where he should take his food truck for the greatest chance of success.<br/>
The aim of this project is to segment area's of Brooklyn's neighborhoods based on the concentration of bars using the Foursquare data.  <br/>We therorize that the best places to take a food truck at night is near areas that have high concentrations of bars.<br/>  This theory is based on two main principals.<br/>
1. Not all bars have a kitchen and sell food, so targeting these areas fills a need for our customers.<br/>
2. People going home from the bars are more likely to buy food to enjoy late at night when many other food options are closed.<br/>

Thus the area's with the highest concentration of bars will have the highest concentrations of potential customers.<br/>




## A description of the data and how it will be used to solve the problem. (15 marks)

We will be using the Foursquare Places API for our data source. https://developer.foursquare.com/docs/api<br/>
From this API we will use the Venue Categories and Venue Location data to map our data on a Folium map.<br/>
From here we will use K-Means clustering to find the ideal locations to send the Food Trucks to.

# III. CODE

### Import necessary Libraries

In [1]:
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize

!conda install -c conda-forge folium=0.5.0 --yes
import folium # plotting library

print('Folium installed')
print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    openssl-1.1.1c             |       h516909a_0         2.1 MB  conda-forge
    geopy-1.20.0               |             py_0          57 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    certifi-2019.9.11          |           py36_0         147 KB  conda-forge
    ca-certificates-2019.9.11  |       hecc5488_0         144 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0         conda-forge
    geopy:           1.20.0-py_0       conda-forge

The following packages will be UPDATED:

    ca-

### Define Foursquare Credentials and Version

##### Make sure that you have created a Foursquare developer account and have your credentials handy

In [2]:
# @hidden
CLIENT_ID = 'CVFZIHHKFIYX5VCCCJJLKGBEKGTCXA2VLQOCMFTXJNFRMIS2' # your Foursquare ID
CLIENT_SECRET = 'IHUGWTRHYEUYIZOP1UAMJKI1FAWCPHI5H22SAHAR1WGF1TXZ' # your Foursquare Secret
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: CVFZIHHKFIYX5VCCCJJLKGBEKGTCXA2VLQOCMFTXJNFRMIS2
CLIENT_SECRET:IHUGWTRHYEUYIZOP1UAMJKI1FAWCPHI5H22SAHAR1WGF1TXZ


#### Let's again assume that you are staying in Brooklyn. So let's start by converting a Brooklyn address to its latitude and longitude coordinates.
In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>foursquare_agent</em>, as shown below.

In [3]:
address = '156 North 4th St, Brooklyn, NY'

geolocator = Nominatim(user_agent="foursquare_agent")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print(latitude, longitude)

40.7157208888889 -73.9589547037037


## 1. Search for a specific venue category
> `https://api.foursquare.com/v2/venues/`**search**`?client_id=`**CLIENT_ID**`&client_secret=`**CLIENT_SECRET**`&ll=`**LATITUDE**`,`**LONGITUDE**`&v=`**VERSION**`&query=`**QUERY**`&radius=`**RADIUS**`&limit=`**LIMIT**

In [4]:
search_query = 'Bar'
radius = 500
print(search_query + ' .... OK!')

Bar .... OK!


#### Define the corresponding URL

In [5]:
url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, search_query, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/search?client_id=CVFZIHHKFIYX5VCCCJJLKGBEKGTCXA2VLQOCMFTXJNFRMIS2&client_secret=IHUGWTRHYEUYIZOP1UAMJKI1FAWCPHI5H22SAHAR1WGF1TXZ&ll=40.7157208888889,-73.9589547037037&v=20180604&query=Bar&radius=500&limit=30'

#### Send the GET Request and examine the results

In [6]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5da078c6c58ed7002cd0b6c9'},
 'response': {'venues': [{'id': '4581734ff964a520653f1fe3',
    'name': 'Surf Bar',
    'location': {'address': '139 N 6th St',
     'crossStreet': 'btwn Bedford & Berry St.',
     'lat': 40.71763628703034,
     'lng': -73.95871390484886,
     'labeledLatLngs': [{'label': 'display',
       'lat': 40.71763628703034,
       'lng': -73.95871390484886}],
     'distance': 214,
     'postalCode': '11249',
     'cc': 'US',
     'city': 'Brooklyn',
     'state': 'NY',
     'country': 'United States',
     'formattedAddress': ['139 N 6th St (btwn Bedford & Berry St.)',
      'Brooklyn, NY 11249',
      'United States']},
    'categories': [{'id': '4bf58dd8d48988d1ce941735',
      'name': 'Seafood Restaurant',
      'pluralName': 'Seafood Restaurants',
      'shortName': 'Seafood',
      'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/seafood_',
       'suffix': '.png'},
      'primary': True}],
    'delivery': {'id

#### Get relevant part of JSON and transform it into a *pandas* dataframe

In [7]:
# assign relevant part of JSON to venues
venues = results['response']['venues']

# tranform venues into a dataframe
dataframe = json_normalize(venues)
dataframe.head()

categories delivery.id  \
0  [{'id': '4bf58dd8d48988d1ce941735', 'name': 'S...      300434   
1  [{'id': '4bf58dd8d48988d152941735', 'name': 'A...         NaN   
2  [{'id': '4bf58dd8d48988d11b941735', 'name': 'P...      316509   
3  [{'id': '4bf58dd8d48988d116941735', 'name': 'B...         NaN   
4  [{'id': '4bf58dd8d48988d110941735', 'name': 'I...      290980   

                delivery.provider.icon.name  \
0  /delivery_provider_seamless_20180129.png   
1                                       NaN   
2  /delivery_provider_seamless_20180129.png   
3                                       NaN   
4  /delivery_provider_seamless_20180129.png   

              delivery.provider.icon.prefix delivery.provider.icon.sizes  \
0  https://fastly.4sqi.net/img/general/cap/                     [40, 50]   
1                                       NaN                          NaN   
2  https://fastly.4sqi.net/img/general/cap/                     [40, 50]   
3                                       NaN                          NaN   
4  https://fastly.4sqi.net/img/general/cap/                     [40, 50]   

  delivery.provider.name                                       delivery.url  \
0               seamless  https://www.seamless.com/menu/surf-bar-139-n-6...   
1                    NaN                                                NaN   
2               seamless  https://www.seamless.com/menu/teddys-bar-and-g...   
3                    NaN                                                NaN   
4               seamless  https://www.seamless.com/menu/doc-wine-bar-83-...   

   hasPerk                        id location.address  ...  \
0    False  4581734ff964a520653f1fe3     139 N 6th St  ...   
1    False  49bb36fcf964a520dc531fe3     291 Grand St  ...   
2    False  4262f880f964a52025211fe3      96 Berry St  ...   
3    False  4af33812f964a520d2eb21e3     318 Grand St  ...   
4    False  4f825769e4b024a26f122b5d      83 N 7th St  ...   

                           location.formattedAddress  \
0  [139 N 6th St (btwn Bedford & Berry St.), Broo...   
1  [291 Grand St (btwn Roebling & Havemeyer St), ...   
2  [96 Berry St (at N. 8th), Brooklyn, NY 11249, ...   
3  [318 Grand St (btwn Havemeyer St. & Marcy Ave....   
4  [83 N 7th St (at Wythe Ave), Brooklyn, NY 1124...   

                             location.labeledLatLngs location.lat  \
0  [{'label': 'display', 'lat': 40.71763628703034...    40.717636   
1  [{'label': 'display', 'lat': 40.71307533802476...    40.713075   
2  [{'label': 'display', 'lat': 40.71920479788144...    40.719205   
3  [{'label': 'display', 'lat': 40.71266183593462...    40.712662   
4  [{'label': 'display', 'lat': 40.71958137689327...    40.719581   

  location.lng  location.neighborhood location.postalCode location.state  \
0   -73.958714                    NaN               11249             NY   
1   -73.957462                    NaN               11211             NY   
2   -73.958431                    NaN               11249             NY   
3   -73.956688                    NaN               11211             NY   
4   -73.960445                    NaN               11249             NY   

                  name    referralId venuePage.id  
0             Surf Bar  v-1570797766     78812387  
1    Caracas Arepa Bar  v-1570797766          NaN  
2  Teddy's Bar & Grill  v-1570797766     78597371  
3      Full Circle Bar  v-1570797766          NaN  
4      D.O.C. Wine Bar  v-1570797766     47216953  

[5 rows x 25 columns]

#### Define information of interest and filter dataframe

In [8]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in dataframe.columns if col.startswith('location.')] + ['id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered['categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
dataframe_filtered.columns = [column.split('.')[-1] for column in dataframe_filtered.columns]

dataframe_filtered

name  \
0                                            Surf Bar   
1                                   Caracas Arepa Bar   
2                                 Teddy's Bar & Grill   
3                                     Full Circle Bar   
4                                     D.O.C. Wine Bar   
5                                         Dim Sum Bar   
6                                         North 4 Bar   
7                                           Dardy Bar   
8                                   Momofuku Milk Bar   
9                                    Woodhul Wine Bar   
10                                        The Bar 245   
11                                    Fresh Kills Bar   
12                                          Baker Bar   
13                          Fabbrica Restaurant & Bar   
14                                 Pinkerton Wine Bar   
15                                   Sort Of Wine Bar   
16                                            Tea Bar   
17                         Brew York City Growler Bar   
18                                            Dry Bar   
19  Mezcaleria La Milagrosa Agave Bar and Listenin...   
20                                      Soft Spot Bar   
21                          La Nonna Ristorante & Bar   
22                              Eddie JR's Sports Bar   
23                                         Lo-Res Bar   
24                            Momofuku Milk Bar Class   
25                                                Bar   
26                  Hammertime Rub Tug And Poetry Bar   
27                                        Squeeze Bar   
28                                 Cadaques Tapas Bar   
29                                      The Trash Bar   

                 categories               address  cc      city  \
0        Seafood Restaurant          139 N 6th St  US  Brooklyn   
1          Arepa Restaurant          291 Grand St  US  Brooklyn   
2                       Pub           96 Berry St  US  Brooklyn   
3                       Bar          318 Grand St  US  Brooklyn   
4        Italian Restaurant           83 N 7th St  US  Brooklyn   
5        Dim Sum Restaurant          167 Grand St  US  Brooklyn   
6                       Bar          160 N 4th St  US  Brooklyn   
7                  Dive Bar     245 S. 1st Street  US  Brooklyn   
8              Dessert Shop  382 Metropolitan Ave  US  Brooklyn   
9                  Wine Bar        644 Driggs Ave  US  Brooklyn   
10                      Bar          245 S 1st St  US  Brooklyn   
11             Cocktail Bar          161 Grand St  US  Brooklyn   
12             Cocktail Bar         200 N 11th St  US  Brooklyn   
13       Italian Restaurant           44 N 6th St  US  New York   
14                 Wine Bar          263 N 6th St  US  Brooklyn   
15                 Wine Bar        639 Driggs Ave  US  Brooklyn   
16                     Café        506 Driggs Ave  US  Brooklyn   
17          Other Nightlife       250 Bedford Ave  US  Brooklyn   
18  Health & Beauty Service          143 Kent Ave  US  New York   
19                      Bar      149 Havemeyer St  US  Brooklyn   
20                      Bar       128 Bedford Ave  US  Brooklyn   
21       Italian Restaurant          184 Kent Ave  US  Brooklyn   
22               Sports Bar          40 Marcy Ave  US  Brooklyn   
23                      Bar  136 Metropolitan Ave  US  Brooklyn   
24           Cooking School        55 Hope Street  US  New York   
25                      Bar          184 Kent Ave  US  New York   
26                     None                   NaN  US       NaN   
27                      Bar       198 Bedford Ave  US  Brooklyn   
28                     None                   NaN  US  Brooklyn   
29                 Dive Bar                   NaN  US  Brooklyn   

          country                       crossStreet  distance  \
0   United States          btwn Bedford & Berry St.       214   
1   United States      btwn Roebling & Havemeyer St       320   
2   Unit

#### Let's visualize the Bars that are nearby

In [9]:
dataframe_filtered.name

0                                              Surf Bar
1                                     Caracas Arepa Bar
2                                   Teddy's Bar & Grill
3                                       Full Circle Bar
4                                       D.O.C. Wine Bar
5                                           Dim Sum Bar
6                                           North 4 Bar
7                                             Dardy Bar
8                                     Momofuku Milk Bar
9                                      Woodhul Wine Bar
10                                          The Bar 245
11                                      Fresh Kills Bar
12                                            Baker Bar
13                            Fabbrica Restaurant & Bar
14                                   Pinkerton Wine Bar
15                                     Sort Of Wine Bar
16                                              Tea Bar
17                           Brew York City Grow

In [10]:
venues_map = folium.Map(location=[latitude, longitude], zoom_start=13) # generate map centred around Williamsburg Brooklyn

# add a red circle marker to represent Williamsburg Brooklyn
folium.features.CircleMarker(
    [latitude, longitude],
    radius=10,
    color='red',
    popup='Brooklyn',
    fill = True,
    fill_color = 'red',
    fill_opacity = 0.6
).add_to(venues_map)

# add the Bars as blue circle markers
for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=5,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(venues_map)

# display map
venues_map

## 2. Explore a Given Venue
> `https://api.foursquare.com/v2/venues/`**VENUE_ID**`?client_id=`**CLIENT_ID**`&client_secret=`**CLIENT_SECRET**`&v=`**VERSION**

In [11]:
venue_id = '49bb36fcf964a520dc531fe3' # ID of Caracas Arepa Bar
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)
url

'https://api.foursquare.com/v2/venues/49bb36fcf964a520dc531fe3?client_id=CVFZIHHKFIYX5VCCCJJLKGBEKGTCXA2VLQOCMFTXJNFRMIS2&client_secret=IHUGWTRHYEUYIZOP1UAMJKI1FAWCPHI5H22SAHAR1WGF1TXZ&v=20180604'

#### Send GET request for result

In [12]:
result = requests.get(url).json()
print(result['response']['venue'].keys())
result['response']['venue']

dict_keys(['id', 'name', 'contact', 'location', 'canonicalUrl', 'categories', 'verified', 'stats', 'url', 'price', 'hasMenu', 'likes', 'dislike', 'ok', 'rating', 'ratingColor', 'ratingSignals', 'menu', 'allowMenuUrlEdit', 'beenHere', 'specials', 'photos', 'reasons', 'description', 'hereNow', 'createdAt', 'tips', 'shortUrl', 'timeZone', 'listed', 'hours', 'popular', 'pageUpdates', 'inbox', 'attributes', 'bestPhoto', 'colors'])


{'id': '49bb36fcf964a520dc531fe3',
 'name': 'Caracas Arepa Bar',
 'contact': {'phone': '7182186050',
  'formattedPhone': '(718) 218-6050',
  'twitter': 'caracasarepabar',
  'instagram': 'caracasarepabar',
  'facebook': '117767434910182',
  'facebookUsername': 'CaracasArepaBar',
  'facebookName': 'Caracas Arepa Bar'},
 'location': {'address': '291 Grand St',
  'crossStreet': 'btwn Roebling & Havemeyer St',
  'lat': 40.71307533802476,
  'lng': -73.9574621701832,
  'labeledLatLngs': [{'label': 'display',
    'lat': 40.71307533802476,
    'lng': -73.9574621701832}],
  'postalCode': '11211',
  'cc': 'US',
  'city': 'Brooklyn',
  'state': 'NY',
  'country': 'United States',
  'formattedAddress': ['291 Grand St (btwn Roebling & Havemeyer St)',
   'Brooklyn, NY 11211',
   'United States']},
 'canonicalUrl': 'https://foursquare.com/v/caracas-arepa-bar/49bb36fcf964a520dc531fe3',
 'categories': [{'id': '4bf58dd8d48988d152941735',
   'name': 'Arepa Restaurant',
   'pluralName': 'Arepa Restaurants'

### B. Get the venue's overall rating

In [13]:
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

9.1


That is a very good rating. Let's check the rating of the second closest Bar.

In [14]:
venue_id = '4581734ff964a520653f1fe3' # ID of Surf Bar
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)

result = requests.get(url).json()
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

7.7


This is not a very good rating.  Let's check the rating of the third closest Bar

In [15]:
venue_id = '4262f880f964a52025211fe3' # ID of Teddy's Bar and Grill
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)

result = requests.get(url).json()
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

7.5


Since this Bar's rating is lower than our first, lets focus on the first Bar.

In [16]:
venue_id = '49bb36fcf964a520dc531fe3' # ID of Caracas Arepa Bar
url = 'https://api.foursquare.com/v2/venues/{}?client_id={}&client_secret={}&v={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION)


result = requests.get(url).json()
try:
    print(result['response']['venue']['rating'])
except:
    print('This venue has not been rated yet.')

9.1


### C. Get the number of tips

In [17]:
result['response']['venue']['tips']['count']

213

### D. Get the venue's tips
> `https://api.foursquare.com/v2/venues/`**VENUE_ID**`/tips?client_id=`**CLIENT_ID**`&client_secret=`**CLIENT_SECRET**`&v=`**VERSION**`&limit=`**LIMIT**

In [18]:
## Caracas Arepa Bar Tips
limit = 15 # set limit to be greater than or equal to the total number of tips
url = 'https://api.foursquare.com/v2/venues/{}/tips?client_id={}&client_secret={}&v={}&limit={}'.format(venue_id, CLIENT_ID, CLIENT_SECRET, VERSION, limit)

results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5da078c85d891b002c39a3aa'},
 'response': {'tips': {'count': 213,
   'items': [{'id': '5907461ef0ca95138496f5de',
     'createdAt': 1493648926,
     'text': 'Delicious arepas, addictive tequenos along with a "we\'re not playing" rum list and a tasty michelada, Caracas is a lasting favorite. The back yard is a fine place in summer.',
     'type': 'user',
     'canonicalUrl': 'https://foursquare.com/item/5907461ef0ca95138496f5de',
     'lang': 'en',
     'likes': {'count': 0, 'groups': []},
     'logView': True,
     'agreeCount': 7,
     'disagreeCount': 0,
     'lastVoteText': 'Upvoted 1 week ago',
     'lastUpvoteTimestamp': 1569628338,
     'todo': {'count': 0},
     'user': {'id': '142228',
      'firstName': 'Marius',
      'lastName': 'Watz',
      'gender': 'male',
      'photo': {'prefix': 'https://fastly.4sqi.net/img/user/',
       'suffix': '/142228-OTAVIHUMKIDNVHV2.jpg'}},
     'authorInteractionType': 'liked'},
    {'id': '576c41bf498e648e

#### Get tips and list of associated features

In [19]:
tips = results['response']['tips']['items']

tip = results['response']['tips']['items'][0]
tip.keys()

dict_keys(['id', 'createdAt', 'text', 'type', 'canonicalUrl', 'lang', 'likes', 'logView', 'agreeCount', 'disagreeCount', 'lastVoteText', 'lastUpvoteTimestamp', 'todo', 'user', 'authorInteractionType'])

#### Format column width and display all tips

In [20]:
pd.set_option('display.max_colwidth', -1)

tips_df = json_normalize(tips) # json normalize tips

# columns to keep
filtered_columns = ['text', 'agreeCount', 'disagreeCount', 'id', 'user.firstName', 'user.lastName', 'user.gender', 'user.id']
tips_filtered = tips_df.loc[:, filtered_columns]

# display tips
tips_filtered

text  \
0  Delicious arepas, addictive tequenos along with a "we're not playing" rum list and a tasty michelada, Caracas is a lasting favorite. The back yard is a fine place in summer.            
1  Consistently delicious - been here 20+ times. Pro tips: Get 3 arepas per 2 people, get the guac and the Yoyos to start with. Get a pitcher of the passionfruit drink to wash it down!!   

   agreeCount  disagreeCount                        id user.firstName  \
0  7           0              5907461ef0ca95138496f5de  Marius          
1  6           0              576c41bf498e648e898aaa59  Leemor          

  user.lastName user.gender   user.id  
0  Watz          male        142228    
1  Yuravlivker   female      20018548

Now remember that because we are using a personal developer account, then we can access only 2 of the restaurant's tips, instead of all 15 tips.

## 3. Search a Foursquare User
> `https://api.foursquare.com/v2/users/`**USER_ID**`?client_id=`**CLIENT_ID**`&client_secret=`**CLIENT_SECRET**`&v=`**VERSION**

### Define URL, send GET request and display features associated with user

In [21]:
user_id = '142228' # user ID with most agree counts and complete profile

url = 'https://api.foursquare.com/v2/users/{}?client_id={}&client_secret={}&v={}'.format(user_id, CLIENT_ID, CLIENT_SECRET, VERSION) # define URL

# send GET request
results = requests.get(url).json()
user_data = results['response']['user']

# display features associated with user
user_data.keys()

dict_keys(['id', 'firstName', 'lastName', 'gender', 'canonicalUrl', 'photo', 'friends', 'tips', 'homeCity', 'bio', 'contact', 'superuser', 'photos', 'type', 'mayorships', 'checkins', 'lists', 'lenses'])

In [22]:
print('First Name: ' + user_data['firstName'])
print('Last Name: ' + user_data['lastName'])
print('Home City: ' + user_data['homeCity'])

First Name: Marius
Last Name: Watz
Home City: Oslo


#### How many tips has this user submitted?

In [23]:
user_data['tips']

{'count': 147}

Wow! So it turns out that Marius is a very active Foursquare user, with more than 147 tips.

### Get User's tips

In [24]:
# define tips URL
url = 'https://api.foursquare.com/v2/users/{}/tips?client_id={}&client_secret={}&v={}&limit={}'.format(user_id, CLIENT_ID, CLIENT_SECRET, VERSION, limit)

# send GET request and get user's tips
results = requests.get(url).json()
tips = results['response']['tips']['items']

# format column width
pd.set_option('display.max_colwidth', -1)

tips_df = json_normalize(tips)

# filter columns
filtered_columns = ['text', 'agreeCount', 'disagreeCount', 'id']
tips_filtered = tips_df.loc[:, filtered_columns]

# display user's tips
tips_filtered

text  \
0   Kongen is the first of the Oslo ferries to switch from gas to exclusively using electric power. So now it’s even easier to take a nap.                                                                 
1   It’s small and not particularly cheap. But hey, it’s local.                                                                                                                                            
2   Delicious coffee, amazing Frühstück. Get the breakfast for 2 and choose 5 types of spread FTW!                                                                                                         
3   Delightful tapas, the black pudding croquetas were amazing.                                                                                                                                            
4   Located in Vulkan, an industrial area redesigned as a hub for food and nightlife, Mathallen offers a mix of restaurants. Also a great spot to see the Akerselva river on your way to Grünerløkka.      
5   Charmingly corny museum dedicated to the voyages of Norwegian hero Thor Heyerdahl. Get a combo ticket and check out the Fram ship next door, too.                                                      
6   An excellent piece of urban design in the middle of the so-so Barcode real estate development, Sørenga is a great place to hang out by the water and even swim if weather allows.                      
7   Delicious cocktail bar with impeccably courteous bartenders. The venue itself is draped in plants, including the herbs they use in the drinks.                                                         
8   Building is from the 1600's, with a great big fireplace and cozy interior. Nice and quiet, it's perfect for a civilized cup of coffee but no guarantees on the food.                                   
9   Cute cafe slash drinking hole on the small idyllic island of Gressholmen. Grab a beer here and combine it with some swimming if the weather permits. (Seasonal.)                                       
10  If you want to cross "see the view in Munch's Scream" off your bucket list, this is the place. Not much else here but the view, but you can walk over to Ekebergparken for a scenic combo trip.        
11  Yes, it's a bus stop, but one where you can cross "the view in Munch's Scream" off your bucket list. Walk uphill and look towards the fjord, then head over to Ekebergparken.                          
12  The commuter ferry to Nesodden is a cheap alternative to a fjord cruise, minus any dining + drink options except a basic on-board kiosk. On a clear day, views are great and the round-trip is <1hr.   
13  Lovely rocky beach that is equally worth visiting in summer and winter. Best reached from Kongsgården bus stop, via a path through the woods. Adventurous? Try nearby Huk nudist beach.                
14  Cute cafe slash drinking hole on the small idyllic island of Gressholmen. Grab a beer here and combine it with some swimming if the weather permits. (Has food, too, but on the pricey side.)          

    agreeCount  disagreeCount                        id  
0   0           0              5d9450569dccf90006f6238e  
1   0           0              5d80d54092667b0006d5619e  
2   0           0              5c24c33cc97f2800395b053a  
3   1           0              5be0c5f9dd12f8002c227040  
4   2           0              5bdc35c06336be002c9873f4  
5   0           0              5bdc350cb9a5a8002c112f4a  
6   0           0              5bdc348a6336be002c983cb0  
7   0           0              5bdb474e32b61d002ca34a3e  
8   0           0              5bdb460fb9a5a8002cef8bcd  
9   0           0              5bdb4459340a58002c89cb82  
10  0           0              5bdb43c4c4df1d002c9d2653  
11  0           0              5bdb2c28c0c0840025fe9679  
12  3           0              5bdb2bbc1af852002cd67bb0  
13  1           0              5bdb2ae0c36588002c4aebca  
14  0           0              5bdb2a6c6336be002c71bc68

#### Let's get the venue for the tip with the greatest number of agree counts

In [25]:
tip_id = '5bdb2bbc1af852002cd67bb0' # tip id

# define URL
url = 'http://api.foursquare.com/v2/tips/{}?client_id={}&client_secret={}&v={}'.format(tip_id, CLIENT_ID, CLIENT_SECRET, VERSION)

# send GET Request and examine results
result = requests.get(url).json()
print(result['response']['tip']['venue']['name'])
print(result['response']['tip']['venue']['location'])

Nesoddbåten
{'lat': 59.910563054562544, 'lng': 10.729425720634632, 'labeledLatLngs': [{'label': 'display', 'lat': 59.910563054562544, 'lng': 10.729425720634632}], 'postalCode': '0256', 'cc': 'NO', 'city': 'Oslo', 'state': 'Oslo', 'country': 'Norge', 'formattedAddress': ['0256 Oslo', 'Norge']}


### Get User's friends

In [26]:
user_friends = json_normalize(user_data['friends']['groups'][0]['items'])
user_friends

bio  \
0  artist, game designer, social network enthusiasts, activist, Homo sapien   
1  tall with short legs.                                                      
2                                                                             
3  What now for man raised by puffins?                                        
4                                                                             
5                                                                             
6                                                                             
7                                                                             
8                                                                             
9  Yes.                                                                       

  firstName  gender            homeCity        id    lastName  \
0  Paba      female  london, uk          161932    B            
1  Scott     male                        2155      Fitzgerald   
2  Simon     male    London, UK          11070394  Sanky        
3  Julian    male    London              513787    Burgess      
4  Mrcl      none    Berlin              29116944  Schwttlck    
5  Joseph    male    Chicago, IL         37839440  Chiocchi     
6  P.        none    Lille               22096120  Box          
7  Chris     female  San Antonio, Texas  23287719  Davila       
8  Valentín  male    Granada, España     6355781   Pedrosa      
9  Joshua    male    Manhattan, NY       39727     Goldberg     

                                      lists.groups  \
0  [{'type': 'created', 'count': 2, 'items': []}]    
1  [{'type': 'created', 'count': 24, 'items': []}]   
2  [{'type': 'created', 'count': 2, 'items': []}]    
3  [{'type': 'created', 'count': 3, 'items': []}]    
4  [{'type': 'created', 'count': 4, 'items': []}]    
5  [{'type': 'created', 'count': 3, 'items': []}]    
6  [{'type': 'created', 'count': 2, 'items': []}]    
7  [{'type': 'created', 'count': 3, 'items': []}]    
8  [{'type': 'created', 'count': 2, 'items': []}]    
9  [{'type': 'created', 'count': 2, 'items': []}]    

                        photo.prefix                    photo.suffix  \
0  https://fastly.4sqi.net/img/user/  /KQODFSEX02EBPRAK.jpg            
1  https://fastly.4sqi.net/img/user/  /-SBP1YXSXHJFIOHVO.jpg           
2  https://fastly.4sqi.net/img/user/  /ZQRTNRXWVMXVMYSB.jpg            
3  https://fastly.4sqi.net/img/user/  /0QNWZERWCIU4HKM5.jpg            
4  https://fastly.4sqi.net/img/user/  /BCZQONIARMRWHT0T.jpg            
5  https://fastly.4sqi.net/img/user/  /FDDUB1ECV3RNOPLG.jpg            
6  https://fastly.4sqi.net/img/user/  /MSYBZV3WMTPUWWO0.png            
7  https://fastly.4sqi.net/img/user/  /23287719-XCBEFZGN5XFAPUO1.jpg   
8  https://fastly.4sqi.net/img/user/  /OET130BD5SXOW3ZU.jpg            
9  https://fastly.4sqi.net/img/user/  /39727_1253904304485.png         

   superuser  tips.count  
0  1.0        70          
1  5.0        59          
2 NaN         0           
3 NaN         27          
4 NaN         3           
5 NaN         0           
6 NaN         0           
7 NaN         5           
8 NaN         2           
9  2.0        8

it appears that Marius has many friends

### Retrieve the User's Profile Image

In [27]:
user_data

{'id': '142228',
 'firstName': 'Marius',
 'lastName': 'Watz',
 'gender': 'male',
 'canonicalUrl': 'https://foursquare.com/user/142228',
 'photo': {'prefix': 'https://fastly.4sqi.net/img/user/',
  'suffix': '/142228-OTAVIHUMKIDNVHV2.jpg'},
 'friends': {'count': 163,
  'groups': [{'type': 'others',
    'name': 'Other friends',
    'count': 163,
    'items': [{'id': '161932',
      'firstName': 'Paba',
      'lastName': 'B',
      'gender': 'female',
      'photo': {'prefix': 'https://fastly.4sqi.net/img/user/',
       'suffix': '/KQODFSEX02EBPRAK.jpg'},
      'tips': {'count': 70},
      'lists': {'groups': [{'type': 'created', 'count': 2, 'items': []}]},
      'homeCity': 'london, uk',
      'bio': 'artist, game designer, social network enthusiasts, activist, Homo sapien',
      'contact': {},
      'superuser': 1},
     {'id': '2155',
      'firstName': 'Scott',
      'lastName': 'Fitzgerald',
      'gender': 'male',
      'photo': {'prefix': 'https://fastly.4sqi.net/img/user/',
      

In [28]:
# 1. grab prefix of photo
# 2. grab suffix of photo
# 3. concatenate them using the image size  
Image(url='https://fastly.4sqi.net/img/user/3000x3000/142228-OTAVIHUMKIDNVHV2.jpg')

## 4. Explore a location
> `https://api.foursquare.com/v2/venues/`**explore**`?client_id=`**CLIENT_ID**`&client_secret=`**CLIENT_SECRET**`&ll=`**LATITUDE**`,`**LONGITUDE**`&v=`**VERSION**`&limit=`**LIMIT**

#### So, you just finished your drink at Caracas Arepa Bar, and are just curious about the popular spots around the bar. In order to explore the area, let's start by getting the latitude and longitude values of Caracas Arepa Barz.

In [29]:
latitude = 40.71307533802476
longitude = -73.9574621701832

In [30]:
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?client_id=CVFZIHHKFIYX5VCCCJJLKGBEKGTCXA2VLQOCMFTXJNFRMIS2&client_secret=IHUGWTRHYEUYIZOP1UAMJKI1FAWCPHI5H22SAHAR1WGF1TXZ&ll=40.71307533802476,-73.9574621701832&v=20180604&radius=500&limit=30'

#### Send GET request and examine results

In [31]:
import requests

In [32]:
results = requests.get(url).json()
'There are {} around Caracas Arepa Bar.'.format(len(results['response']['groups'][0]['items']))

'There are 30 around Caracas Arepa Bar.'

#### Get relevant part of JSON

In [33]:
items = results['response']['groups'][0]['items']
items[0]

{'reasons': {'count': 0,
  'items': [{'summary': 'This spot is popular',
    'type': 'general',
    'reasonName': 'globalInteractionReason'}]},
 'venue': {'id': '4e4ea8b5aeb70f12849528ed',
  'name': 'St. Mazie',
  'location': {'address': '345 Grand St',
   'crossStreet': 'btwn Marcy Ave & Havemeyer St',
   'lat': 40.712446182313414,
   'lng': -73.95595408442468,
   'labeledLatLngs': [{'label': 'display',
     'lat': 40.712446182313414,
     'lng': -73.95595408442468}],
   'distance': 145,
   'postalCode': '11211',
   'cc': 'US',
   'city': 'Brooklyn',
   'state': 'NY',
   'country': 'United States',
   'formattedAddress': ['345 Grand St (btwn Marcy Ave & Havemeyer St)',
    'Brooklyn, NY 11211',
    'United States']},
  'categories': [{'id': '4bf58dd8d48988d11e941735',
    'name': 'Cocktail Bar',
    'pluralName': 'Cocktail Bars',
    'shortName': 'Cocktail',
    'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/nightlife/cocktails_',
     'suffix': '.png'},
    'primary': Tru

#### Process JSON and convert it to a clean dataframe

In [34]:
dataframe = json_normalize(items) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories'] + [col for col in dataframe.columns if col.startswith('venue.location.')] + ['venue.id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# filter the category for each row
dataframe_filtered['venue.categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean columns
dataframe_filtered.columns = [col.split('.')[-1] for col in dataframe_filtered.columns]

dataframe_filtered.head(10)

name           categories               address  cc  \
0  St. Mazie               Cocktail Bar         345 Grand St          US   
1  East River Pilates      Pilates Studio       235 S 1st St          US   
2  L'Industrie Pizzeria    Pizza Place          254 S 2nd St          US   
3  St. Anselm              American Restaurant  355 Metropolitan Ave  US   
4  Emmy Squared            Pizza Place          364 Grand St          US   
5  Twenty Sided Store      Toy / Game Store     362 Grand St          US   
6  Have & Meyer Chatteria  Wine Bar             103 Havermeyer St     US   
7  The Four Horsemen       Wine Bar             295 Grand St          US   
8  Bozu                    Sushi Restaurant     296 Grand St          US   
9  Caracas Arepa Bar       Arepa Restaurant     291 Grand St          US   

       city        country                    crossStreet  distance  \
0  Brooklyn  United States  btwn Marcy Ave & Havemeyer St  145        
1  Brooklyn  United States  at Roebling street             89         
2  Brooklyn  United States  Havemeyer                      157        
3  Brooklyn  United States  btwn Havemeyer & Roebling St   172        
4  Brooklyn  United States  at Marcy Ave                   179        
5  Brooklyn  United States  at Marcy St                    190        
6  Brooklyn  United States  Hope St                        88         
7  Brooklyn  United States  btwn Havemeyer & Roebling St   15         
8  Brooklyn  United States  btwn Roebling & Havemeyer St   11         
9  Brooklyn  United States  btwn Roebling & Havemeyer St   0          

                                                                           formattedAddress  \
0  [345 Grand St (btwn Marcy Ave & Havemeyer St), Brooklyn, NY 11211, United States]          
1  [235 S 1st St (at Roebling street), Brooklyn, NY 11211, United States]                     
2  [254 S 2nd St (Havemeyer), Brooklyn, NY 11211, United States]                              
3  [355 Metropolitan Ave (btwn Havemeyer & Roebling St), Brooklyn, NY 11211, United States]   
4  [364 Grand St (at Marcy Ave), Brooklyn, NY 11211, United States]                           
5  [362 Grand St (at Marcy St), Brooklyn, NY 11211, United States]                            
6  [103 Havermeyer St (Hope St), Brooklyn, NY 11211, United States]                           
7  [295 Grand St (btwn Havemeyer & Roebling St), Brooklyn, NY 11211, United States]           
8  [296 Grand St (btwn Roebling & Havemeyer St), Brooklyn, NY 11211, United States]           
9  [291 Grand St (btwn Roebling & Havemeyer St), Brooklyn, NY 11211, United States]           

                                                                 labeledLatLngs  \
0  [{'label': 'display', 'lat': 40.712446182313414, 'lng': -73.95595408442468}]   
1  [{'label': 'display', 'lat': 40.71289117042526, 'lng': -73.95850006447007}]    
2  [{'label': 'display', 'lat': 40.7116752206617, 'lng': -73.95769662100513}]     
3  [{'label': 'display', 'lat': 40.71414471294114, 'lng': -73.95598405757656}]    
4  [{'label': 'display', 'lat': 40.712166, 'lng': -73.955705}]                    
5  [{'label': 'display', 'lat': 40.712268329253185, 'lng': -73.9554761947898}]    
6  [{'label': 'display', 'lat': 40.71330486799995, 'lng': -73.95646386109051}]    
7  [{'label': 'display', 'lat': 40.71297062908573, 'lng': -73.9573489459191}]     
8  [{'label': 'display', 'lat': 40.71297626571757, 'lng': -73.95747790354523}]    
9  [{'label': 'display', 'lat': 40.71307533802476, 'lng': -73.9574621701832}]     

         lat        lng neighborhood postalCode state  \
0  40.712446 -73.955954  NaN          11211      NY     
1  40.712891 -73.958500  NaN          11211      NY     
2  40.711675 -73.957697  NaN          11211      NY     
3  40.714145 -73.955984  NaN          11211      NY     
4  40.712166 -73.955705  NaN          11211      NY     
5  40.712268 -73.955476  NaN          11211      NY     
6  40.713305 -73.956464  NaN          11211

#### Let's visualize these items on the map around our location

In [35]:
venues_map = folium.Map(location=[latitude, longitude], zoom_start=15) # generate map centred around Ecco


# add Ecco as a red circle mark
folium.features.CircleMarker(
    [latitude, longitude],
    radius=10,
    popup='Ecco',
    fill=True,
    color='red',
    fill_color='red',
    fill_opacity=0.6
    ).add_to(venues_map)


# add popular spots to the map as blue circle markers
for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        fill=True,
        color='blue',
        fill_color='blue',
        fill_opacity=0.6
        ).add_to(venues_map)

# display map
venues_map

## 5. Explore Trending Venues
> `https://api.foursquare.com/v2/venues/`**trending**`?client_id=`**CLIENT_ID**`&client_secret=`**CLIENT_SECRET**`&ll=`**LATITUDE**`,`**LONGITUDE**`&v=`**VERSION**

#### Now, instead of simply exploring the area around Caracas Arepa Bar, you are interested in knowing the venues that are trending at the time when people are out at the bars, meaning the places with the highest foot traffic. So let's do that and get the trending venues around Caracas Arepa Bar.

In [36]:
# define URL
url = 'https://api.foursquare.com/v2/venues/trending?client_id={}&client_secret={}&ll={},{}&v={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION)

# send GET request and get trending venues
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5da078ca598e64002c23853a'},
 'response': {'venues': []}}

### Check if any venues are trending at this time

In [37]:
if len(results['response']['venues']) == 0:
    trending_venues_df = 'No trending venues are available at the moment!'
    
else:
    trending_venues = results['response']['venues']
    trending_venues_df = json_normalize(trending_venues)

    # filter columns
    columns_filtered = ['name', 'categories'] + ['location.distance', 'location.city', 'location.postalCode', 'location.state', 'location.country', 'location.lat', 'location.lng']
    trending_venues_df = trending_venues_df.loc[:, columns_filtered]

    # filter the category for each row
    trending_venues_df['categories'] = trending_venues_df.apply(get_category_type, axis=1)

In [38]:
# display trending venues
trending_venues_df

'No trending venues are available at the moment!'

Now, depending on when you run the above code, you might get different venues since the venues with the highest foot traffic are fetched live. 

### Visualize trending venues

In [39]:
if len(results['response']['venues']) == 0:
    venues_map = 'Cannot generate visual as no trending venues are available at the moment!'

else:
    venues_map = folium.Map(location=[latitude, longitude], zoom_start=15) # generate map centred around Ecco


    # add Caracas Arepa Bar as a red circle mark
    folium.features.CircleMarker(
        [latitude, longitude],
        radius=10,
        popup='Caracas Arepa Bar',
        fill=True,
        color='red',
        fill_color='red',
        fill_opacity=0.6
    ).add_to(venues_map)


    # add the trending venues as blue circle markers
    for lat, lng, label in zip(trending_venues_df['location.lat'], trending_venues_df['location.lng'], trending_venues_df['name']):
        folium.features.CircleMarker(
            [lat, lng],
            radius=5,
            poup=label,
            fill=True,
            color='blue',
            fill_color='blue',
            fill_opacity=0.6
        ).add_to(venues_map)

In [40]:
# display map
venues_map

'Cannot generate visual as no trending venues are available at the moment!'

<h1 align=center><font size = 5>Segmenting and Clustering Neighborhoods in Brooklyn</font></h1>

## Introduction

In this lab, you will learn how to convert addresses into their equivalent latitude and longitude values. Also, you will use the Foursquare API to explore neighborhoods in New York City. You will use the **explore** function to get the most common venue categories in each neighborhood, and then use this feature to group the neighborhoods into clusters. You will use the *k*-means clustering algorithm to complete this task. Finally, you will use the Folium library to visualize the neighborhoods in New York City and their emerging clusters.

## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>

1. <a href="#item1">Download and Explore Dataset</a>

2. <a href="#item2">Explore Neighborhoods in Brooklyn</a>

3. <a href="#item3">Analyze Each Neighborhood</a>

4. <a href="#item4">Cluster Neighborhoods</a>

5. <a href="#item5">Examine Clusters</a>    
</font>
</div>

Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [42]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


## 1. Download and Explore Dataset

Neighborhood has a total of 5 boroughs and 306 neighborhoods. In order to segement the neighborhoods and explore them, we will essentially need a dataset that contains the 5 boroughs and the neighborhoods that exist in each borough as well as the the latitude and logitude coordinates of each neighborhood. 

Luckily, this dataset exists for free on the web. Feel free to try to find this dataset on your own, but here is the link to the dataset: https://geo.nyu.edu/catalog/nyu_2451_34572

In [43]:
!wget -q -O 'newyork_data.json' https://cocl.us/new_york_dataset
print('Data downloaded!')

Data downloaded!


#### Load and explore the data

Next, let's load the data.

In [46]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

In [47]:
newyork_data

{'type': 'FeatureCollection',
 'totalFeatures': 306,
 'features': [{'type': 'Feature',
   'id': 'nyu_2451_34572.1',
   'geometry': {'type': 'Point',
    'coordinates': [-73.84720052054902, 40.89470517661]},
   'geometry_name': 'geom',
   'properties': {'name': 'Wakefield',
    'stacked': 1,
    'annoline1': 'Wakefield',
    'annoline2': None,
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.84720052054902,
     40.89470517661,
     -73.84720052054902,
     40.89470517661]}},
  {'type': 'Feature',
   'id': 'nyu_2451_34572.2',
   'geometry': {'type': 'Point',
    'coordinates': [-73.82993910812398, 40.87429419303012]},
   'geometry_name': 'geom',
   'properties': {'name': 'Co-op City',
    'stacked': 2,
    'annoline1': 'Co-op',
    'annoline2': 'City',
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.82993910812398,
     40.87429419303012,
     -73.82993910812398,
     40.87429419303012]}},
  {'type': 'Feature',
 

In [48]:
neighborhoods_data = newyork_data['features']

In [49]:
neighborhoods_data[0]

{'type': 'Feature',
 'id': 'nyu_2451_34572.1',
 'geometry': {'type': 'Point',
  'coordinates': [-73.84720052054902, 40.89470517661]},
 'geometry_name': 'geom',
 'properties': {'name': 'Wakefield',
  'stacked': 1,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'annoangle': 0.0,
  'borough': 'Bronx',
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661]}}

#### Tranform the data into a *pandas* dataframe

The next task is essentially transforming this data of nested Python dictionaries into a *pandas* dataframe. So let's start by creating an empty dataframe.

In [50]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)

Take a look at the empty dataframe to confirm that the columns are as intended.

In [51]:
neighborhoods

Empty DataFrame
Columns: [Borough, Neighborhood, Latitude, Longitude]
Index: []

Then let's loop through the data and fill the dataframe one row at a time.

In [53]:
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

Quickly examine the resulting dataframe.

In [54]:
neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0  Bronx   Wakefield    40.894705 -73.847201
1  Bronx   Co-op City   40.874294 -73.829939
2  Bronx   Eastchester  40.887556 -73.827806
3  Bronx   Fieldston    40.895437 -73.905643
4  Bronx   Riverdale    40.890834 -73.912585

And make sure that the dataset has all 5 boroughs and 306 neighborhoods.

In [56]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 5 boroughs and 612 neighborhoods.


#### Use geopy library to get the latitude and longitude values of New York City.

In [69]:
address = 'Brooklyn, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Brooklyn are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Brooklyn are 40.6501038, -73.9495823.


#### Create a map of New York with neighborhoods superimposed on top.

In [58]:
# create map of New York using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

**Folium** is a great visualization library. Feel free to zoom into the above map, and click on each circle mark to reveal the name of the neighborhood and its respective borough.

However, for illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods in Manhattan. So let's slice the original dataframe and create a new dataframe of the Manhattan data.

In [65]:
manhattan_data = neighborhoods[neighborhoods['Borough'] == 'Brooklyn'].reset_index(drop=True)
manhattan_data.head()

Borough Neighborhood   Latitude  Longitude
0  Brooklyn  Bay Ridge    40.625801 -74.030621
1  Brooklyn  Bensonhurst  40.611009 -73.995180
2  Brooklyn  Sunset Park  40.645103 -74.010316
3  Brooklyn  Greenpoint   40.730201 -73.954241
4  Brooklyn  Gravesend    40.595260 -73.973471

Let's get the geographical coordinates of Manhattan.

In [66]:
address = 'Brooklyn, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Manhattan are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Manhattan are 40.6501038, -73.9495823.


As we did with all of New York City, let's visualizat Manhattan the neighborhoods in it.

In [67]:
# create map of Manhattan using latitude and longitude values
map_manhattan = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(manhattan_data['Latitude'], manhattan_data['Longitude'], manhattan_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_manhattan)  
    
map_manhattan

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [70]:
CLIENT_ID = 'CVFZIHHKFIYX5VCCCJJLKGBEKGTCXA2VLQOCMFTXJNFRMIS2' # your Foursquare ID
CLIENT_SECRET = 'IHUGWTRHYEUYIZOP1UAMJKI1FAWCPHI5H22SAHAR1WGF1TXZ' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: CVFZIHHKFIYX5VCCCJJLKGBEKGTCXA2VLQOCMFTXJNFRMIS2
CLIENT_SECRET:IHUGWTRHYEUYIZOP1UAMJKI1FAWCPHI5H22SAHAR1WGF1TXZ


#### Let's explore the first neighborhood in our dataframe.
Get the neighborhood's name.

In [71]:
manhattan_data.loc[0, 'Neighborhood']

'Bay Ridge'

Get the neighborhood's latitude and longitude values.

In [72]:
neighborhood_latitude = manhattan_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = manhattan_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = manhattan_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Bay Ridge are 40.625801065010656, -74.03062069353813.


#### Now, let's get the top 100 venues that are in Bay Ridge within a radius of 500 meters.

In [73]:
# type your answer here
LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)

url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=CVFZIHHKFIYX5VCCCJJLKGBEKGTCXA2VLQOCMFTXJNFRMIS2&client_secret=IHUGWTRHYEUYIZOP1UAMJKI1FAWCPHI5H22SAHAR1WGF1TXZ&v=20180605&ll=40.625801065010656,-74.03062069353813&radius=500&limit=100'

Send the GET request and examine the resutls

In [74]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5da083fdfd16bb002c1f14e6'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': '$-$$$$', 'key': 'price'},
    {'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Bay Ridge',
  'headerFullLocation': 'Bay Ridge, Brooklyn',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 86,
  'suggestedBounds': {'ne': {'lat': 40.63030106951066,
    'lng': -74.02470273356597},
   'sw': {'lat': 40.62130106051065, 'lng': -74.03653865351028}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4b895827f964a5206c2d32e3',
       'name': 'Pilo Arts Day Spa and Salon',
       'location': {'address': '8412 3rd Ave',
        'lat': 40.62474788273414,
        'lng': -74.03059056940135,
        'labeledLatL

From the Foursquare lab in the previous module, we know that all the information is in the *items* key. Before we proceed, let's borrow the **get_category_type** function from the Foursquare lab.

In [75]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a *pandas* dataframe.

In [76]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name      categories        lat        lng
0  Pilo Arts Day Spa and Salon  Spa             40.624748 -74.030591
1  Bagel Boy                    Bagel Shop      40.627896 -74.029335
2  Cocoa Grinder                Juice Bar       40.623967 -74.030863
3  Pegasus Cafe                 Breakfast Spot  40.623168 -74.031186
4  Ho' Brah Taco Joint          Taco Place      40.622960 -74.031371

And how many venues were returned by Foursquare?

In [77]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

86 venues were returned by Foursquare.


## 2. Explore Neighborhoods in Brooklyn

#### Let's create a function to repeat the same process to all the neighborhoods in Brooklyn

In [80]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called *manhattan_venues*.

In [81]:
manhattan_venues = getNearbyVenues(names=manhattan_data['Neighborhood'],
                                   latitudes=manhattan_data['Latitude'],
                                   longitudes=manhattan_data['Longitude']
                                  )

Bay Ridge
Bensonhurst
Sunset Park
Greenpoint
Gravesend
Brighton Beach
Sheepshead Bay
Manhattan Terrace
Flatbush
Crown Heights
East Flatbush
Kensington
Windsor Terrace
Prospect Heights
Brownsville
Williamsburg
Bushwick
Bedford Stuyvesant
Brooklyn Heights
Cobble Hill
Carroll Gardens
Red Hook
Gowanus
Fort Greene
Park Slope
Cypress Hills
East New York
Starrett City
Canarsie
Flatlands
Mill Island
Manhattan Beach
Coney Island
Bath Beach
Borough Park
Dyker Heights
Gerritsen Beach
Marine Park
Clinton Hill
Sea Gate
Downtown
Boerum Hill
Prospect Lefferts Gardens
Ocean Hill
City Line
Bergen Beach
Midwood
Prospect Park South
Georgetown
East Williamsburg
North Side
South Side
Ocean Parkway
Fort Hamilton
Ditmas Park
Wingate
Rugby
Remsen Village
New Lots
Paerdegat Basin
Mill Basin
Fulton Ferry
Vinegar Hill
Weeksville
Broadway Junction
Dumbo
Homecrest
Highland Park
Madison
Erasmus
Bay Ridge
Bensonhurst
Sunset Park
Greenpoint
Gravesend
Brighton Beach
Sheepshead Bay
Manhattan Terrace
Flatbush
Crown Heig

#### Let's check the size of the resulting dataframe

In [82]:
print(manhattan_venues.shape)
manhattan_venues.head()

(5696, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Bay Ridge    40.625801             -74.030621                
1  Bay Ridge    40.625801             -74.030621                
2  Bay Ridge    40.625801             -74.030621                
3  Bay Ridge    40.625801             -74.030621                
4  Bay Ridge    40.625801             -74.030621                

                         Venue  Venue Latitude  Venue Longitude  \
0  Pilo Arts Day Spa and Salon  40.624748      -74.030591         
1  Bagel Boy                    40.627896      -74.029335         
2  Cocoa Grinder                40.623967      -74.030863         
3  Pegasus Cafe                 40.623168      -74.031186         
4  Ho' Brah Taco Joint          40.622960      -74.031371         

   Venue Category  
0  Spa             
1  Bagel Shop      
2  Juice Bar       
3  Breakfast Spot  
4  Taco Place

Let's check how many venues were returned for each neighborhood

In [83]:
manhattan_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  \
Neighborhood                                                               
Bath Beach                 98                     98                       
Bay Ridge                  172                    172                      
Bedford Stuyvesant         52                     52                       
Bensonhurst                74                     74                       
Bergen Beach               12                     12                       
Boerum Hill                182                    182                      
Borough Park               44                     44                       
Brighton Beach             88                     88                       
Broadway Junction          28                     28                       
Brooklyn Heights           200                    200                      
Brownsville                36                     36                       
Bushwick                   150                    150                      
Canarsie                   14                     14                       
Carroll Gardens            200                    200                      
City Line                  62                     62                       
Clinton Hill               176                    176                      
Cobble Hill                200                    200                      
Coney Island               38                     38                       
Crown Heights              38                     38                       
Cypress Hills              60                     60                       
Ditmas Park                104                    104                      
Downtown                   200                    200                      
Dumbo                      176                    176                      
Dyker Heights              10                     10                       
East Flatbush              20                     20                       
East New York              20                     20                       
East Williamsburg          140                    140                      
Erasmus                    38                     38                       
Flatbush                   48                     48                       
Flatlands                  38                     38                       
Fort Greene                162                    162                      
Fort Hamilton              132                    132                      
Fulton Ferry               126                    126                      
Georgetown                 64                     64                       
Gerritsen Beach            40                     40                       
Gowanus                    138                    138                      
Gravesend                  60                     60                       
Greenpoint                 200                    200                      
Highland Park              28                     28                       
Homecrest                  88                     88                       
Kensington                 76                     76                       
Madison                    24                     24                       
Manhattan Beach            18                     18                       
Manhattan Terrace          52                     52                       
Marine Park                16                     16                       
Midwood                    24                     24                       
Mill Basin                 70                     70                       
Mill Island                2                      2                        
New Lots                   44                     44                       
North Side                 200                    200                      
Ocean Hill                 60                     60                       

#### Let's find out how many unique categories can be curated from all the returned venues

In [84]:
print('There are {} uniques categories.'.format(len(manhattan_venues['Venue Category'].unique())))

There are 287 uniques categories.


## 3. Analyze Each Neighborhood

In [85]:
# one hot encoding
manhattan_onehot = pd.get_dummies(manhattan_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
manhattan_onehot['Neighborhood'] = manhattan_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [manhattan_onehot.columns[-1]] + list(manhattan_onehot.columns[:-1])
manhattan_onehot = manhattan_onehot[fixed_columns]

manhattan_onehot.head()

Yoga Studio  Accessories Store  Adult Boutique  Airport Terminal  \
0  0            0                  0               0                  
1  0            0                  0               0                  
2  0            0                  0               0                  
3  0            0                  0               0                  
4  0            0                  0               0                  

   American Restaurant  Antique Shop  Arepa Restaurant  \
0  0                    0             0                  
1  0                    0             0                  
2  0                    0             0                  
3  0                    0             0                  
4  0                    0             0                  

   Argentinian Restaurant  Art Gallery  Arts & Crafts Store  \
0  0                       0            0                     
1  0                       0            0                     
2  0                       0            0                     
3  0                       0            0                     
4  0                       0            0                     

   Arts & Entertainment  Asian Restaurant  Athletics & Sports  Auto Garage  \
0  0                     0                 0                   0             
1  0                     0                 0                   0             
2  0                     0                 0                   0             
3  0                     0                 0                   0             
4  0                     0                 0                   0             

   BBQ Joint  Bagel Shop  Bakery  Bank  Bar  Baseball Field  Baseball Stadium  \
0  0          0           0       0     0    0               0                  
1  0          1           0       0     0    0               0                  
2  0          0           0       0     0    0               0                  
3  0          0           0       0     0    0               0                  
4  0          0           0       0     0    0               0                  

   Basketball Court  Beach  Beer Bar  Beer Garden  Beer Store  Big Box Store  \
0  0                 0      0         0            0           0               
1  0                 0      0         0            0           0               
2  0                 0      0         0            0           0               
3  0                 0      0         0            0           0               
4  0                 0      0         0            0           0               

   Bike Rental / Bike Share  Bike Shop  Bistro  Board Shop  Boat or Ferry  \
0  0                         0          0       0           0               
1  0                         0          0       0           0               
2  0                         0          0       0           0               
3  0                         0          0       0           0               
4  0                         0          0       0           0               

   Bookstore  Boutique  Boxing Gym  Breakfast Spot  Brewery  Bridal Shop  \
0  0          0         0           0               0        0             
1  0          0         0           0               0        0             
2  0          0         0           0               0        0             
3  0          0         0           1               0        0             
4  0          0         0           0               0        0             

   Bridge  Bubble Tea Shop  Buffet  Burger Joint  Burrito Place  Bus Line  \
0  0       0                0       0             0              0          
1  0       0                0       0             0              0          
2  0       0                0       0             0              0          
3  0       0                0       0             0              0          
4  0       0                0       0             0              0          

   Bus Station  Bus Stop  Bu

And let's examine the new dataframe size.

In [86]:
manhattan_onehot.shape

(5696, 287)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [87]:
manhattan_grouped = manhattan_onehot.groupby('Neighborhood').mean().reset_index()
manhattan_grouped

Neighborhood  Yoga Studio  Accessories Store  Adult Boutique  \
0   Bath Beach                 0.000000     0.000000           0.00             
1   Bay Ridge                  0.011628     0.000000           0.00             
2   Bedford Stuyvesant         0.000000     0.000000           0.00             
3   Bensonhurst                0.000000     0.000000           0.00             
4   Bergen Beach               0.000000     0.000000           0.00             
5   Boerum Hill                0.021978     0.000000           0.00             
6   Borough Park               0.000000     0.000000           0.00             
7   Brighton Beach             0.000000     0.000000           0.00             
8   Broadway Junction          0.000000     0.000000           0.00             
9   Brooklyn Heights           0.050000     0.000000           0.00             
10  Brownsville                0.000000     0.000000           0.00             
11  Bushwick                   0.000000     0.000000           0.00             
12  Canarsie                   0.000000     0.000000           0.00             
13  Carroll Gardens            0.000000     0.000000           0.00             
14  City Line                  0.000000     0.000000           0.00             
15  Clinton Hill               0.034091     0.000000           0.00             
16  Cobble Hill                0.030000     0.000000           0.00             
17  Coney Island               0.000000     0.000000           0.00             
18  Crown Heights              0.000000     0.000000           0.00             
19  Cypress Hills              0.000000     0.000000           0.00             
20  Ditmas Park                0.019231     0.000000           0.00             
21  Downtown                   0.010000     0.000000           0.00             
22  Dumbo                      0.022727     0.000000           0.00             
23  Dyker Heights              0.000000     0.000000           0.00             
24  East Flatbush              0.000000     0.000000           0.00             
25  East New York              0.000000     0.000000           0.00             
26  East Williamsburg          0.000000     0.000000           0.00             
27  Erasmus                    0.052632     0.000000           0.00             
28  Flatbush                   0.000000     0.000000           0.00             
29  Flatlands                  0.000000     0.000000           0.00             
30  Fort Greene                0.024691     0.000000           0.00             
31  Fort Hamilton              0.000000     0.000000           0.00             
32  Fulton Ferry               0.000000     0.000000           0.00             
33  Georgetown                 0.000000     0.000000           0.00             
34  Gerritsen Beach            0.000000     0.000000           0.00             
35  Gowanus                    0.014493     0.000000           0.00             
36  Gravesend                  0.000000     0.000000           0.00             
37  Greenpoint                 0.030000     0.000000           0.00             
38  Highland Park              0.000000     0.000000           0.00             
39  Homecrest                  0.000000     0.000000           0.00             
40  Kensington                 0.000000     0.000000           0.00             
41  Madison                    0.000000     0.000000           0.00             
42  Manhattan Beach            0.000000     0.000000           0.00             
43  Manhattan Terrace          0.000000     0.000000           0.00             
44  Marine Park                0.000000     0.000000           0.00             
45  Midwood                    0.000000     0.000000           0.00             
46  Mill Basin                 0.000000     0.000000           0.00             
47  Mill Island                0.000000     0.000000           0.00             
48  New Lots                   0.000000     0.00

#### Let's confirm the new size

In [88]:
manhattan_grouped.shape

(70, 287)

#### Let's print each neighborhood along with the top 5 most common venues

In [89]:
num_top_venues = 5

for hood in manhattan_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = manhattan_grouped[manhattan_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Bath Beach----
                venue  freq
0  Chinese Restaurant  0.08
1  Pharmacy            0.06
2  Bubble Tea Shop     0.04
3  Kids Store          0.04
4  Pizza Place         0.04


----Bay Ridge----
                 venue  freq
0  Spa                  0.07
1  Italian Restaurant   0.07
2  Pizza Place          0.05
3  Chinese Restaurant   0.03
4  American Restaurant  0.03


----Bedford Stuyvesant----
           venue  freq
0  Coffee Shop    0.08
1  Pizza Place    0.08
2  Café           0.08
3  Bar            0.08
4  Deli / Bodega  0.08


----Bensonhurst----
                venue  freq
0  Chinese Restaurant  0.11
1  Donut Shop          0.05
2  Grocery Store       0.05
3  Bakery              0.05
4  Spa                 0.05


----Bergen Beach----
                venue  freq
0  Harbor / Marina     0.33
1  Baseball Field      0.17
2  Athletics & Sports  0.17
3  Playground          0.17
4  Donut Shop          0.17


----Boerum Hill----
                    venue  freq
0  Coffee Shop   

#### Let's put that into a *pandas* dataframe
First, let's write a function to sort the venues in descending order.

In [90]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [92]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = manhattan_grouped['Neighborhood']

for ind in np.arange(manhattan_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(manhattan_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0  Bath Beach          Chinese Restaurant    Pharmacy               
1  Bay Ridge           Spa                   Italian Restaurant     
2  Bedford Stuyvesant  Deli / Bodega         Pizza Place            
3  Bensonhurst         Chinese Restaurant    Ice Cream Shop         
4  Bergen Beach        Harbor / Marina       Playground             

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0  Kids Store            Bubble Tea Shop       Donut Shop             
1  Pizza Place           American Restaurant   Bar                    
2  Coffee Shop           Café                  Bar                    
3  Donut Shop            Sushi Restaurant      Spa                    
4  Baseball Field        Athletics & Sports    Donut Shop             

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0  Pizza Place             Italian Restaurant    Fast Food Restaurant   
1  Greek Restaurant        Chinese Restaurant    Sushi Restaurant       
2  Thrift / Vintage Store  Park                  Cocktail Bar           
3  Grocery Store           Bakery                Pizza Place            
4  Factory                 Falafel Restaurant    Farm                   

     9th Most Common Venue 10th Most Common Venue  
0  Ice Cream Shop           Mobile Phone Shop      
1  Thai Restaurant          Hookah Bar             
2  New American Restaurant  Bus Stop               
3  Park                     Noodle House           
4  Farmers Market           Fast Food Restaurant

## 4. Cluster Neighborhoods
Run *k*-means to cluster the neighborhood into 5 clusters.

In [93]:
# set number of clusters
kclusters = 5

manhattan_grouped_clustering = manhattan_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(manhattan_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 4, 4, 1, 4, 4, 4, 4, 4, 4], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [94]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

manhattan_merged = manhattan_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
manhattan_merged = manhattan_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

manhattan_merged.head() # check the last columns!

Borough Neighborhood   Latitude  Longitude  Cluster Labels  \
0  Brooklyn  Bay Ridge    40.625801 -74.030621  4                
1  Brooklyn  Bensonhurst  40.611009 -73.995180  1                
2  Brooklyn  Sunset Park  40.645103 -74.010316  4                
3  Brooklyn  Greenpoint   40.730201 -73.954241  4                
4  Brooklyn  Gravesend    40.595260 -73.973471  1                

       1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0  Spa                        Italian Restaurant    Pizza Place            
1  Chinese Restaurant         Ice Cream Shop        Donut Shop             
2  Latin American Restaurant  Bakery                Bank                   
3  Bar                        Coffee Shop           Pizza Place            
4  Italian Restaurant         Chinese Restaurant    Bus Station            

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0  American Restaurant   Bar                   Greek Restaurant       
1  Sushi Restaurant      Spa                   Grocery Store          
2  Pizza Place           Mexican Restaurant    Ice Cream Shop         
3  Cocktail Bar          Yoga Studio           Record Shop            
4  Pizza Place           Lounge                Martial Arts Dojo      

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0  Chinese Restaurant    Sushi Restaurant      Thai Restaurant        
1  Bakery                Pizza Place           Park                   
2  Pharmacy              Deli / Bodega         Mobile Phone Shop      
3  Mexican Restaurant    Sushi Restaurant      French Restaurant      
4  Bakery                Gym                   Liquor Store           

  10th Most Common Venue  
0  Hookah Bar             
1  Noodle House           
2  Gym                    
3  Café                   
4  Baseball Field

Finally, let's visualize the resulting clusters

In [95]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(manhattan_merged['Latitude'], manhattan_merged['Longitude'], manhattan_merged['Neighborhood'], manhattan_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 5. Examine Clusters
Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.

#### Cluster 1

In [96]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 0, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
7    Manhattan Terrace  Pizza Place           Ice Cream Shop         
46   Midwood            Pizza Place           Ice Cream Shop         
77   Manhattan Terrace  Pizza Place           Ice Cream Shop         
116  Midwood            Pizza Place           Ice Cream Shop         

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
7    Cosmetics Shop        Donut Shop            Grocery Store          
46   Pharmacy              Candy Store           Convenience Store      
77   Cosmetics Shop        Donut Shop            Grocery Store          
116  Pharmacy              Candy Store           Convenience Store      

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
7    Convenience Store     Coffee Shop           Organic Grocery        
46   Video Game Store      Italian Restaurant    Bakery                 
77   Convenience Store     Coffee Shop           Organic Grocery        
116  Video Game Store      Italian Restaurant    Bakery                 

    9th Most Common Venue 10th Most Common Venue  
7    Chinese Restaurant    Bus Station            
46   Bagel Shop            Field                  
77   Chinese Restaurant    Bus Station            
116  Bagel Shop            Field

#### Cluster 2

In [97]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 1, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood      1st Most Common Venue  \
1    Bensonhurst          Chinese Restaurant          
4    Gravesend            Italian Restaurant          
10   East Flatbush        Food & Drink Shop           
25   Cypress Hills        Latin American Restaurant   
27   Starrett City        Pizza Place                 
28   Canarsie             IT Services                 
29   Flatlands            Pharmacy                    
33   Bath Beach           Chinese Restaurant          
39   Sea Gate             Spa                         
44   City Line            Donut Shop                  
47   Prospect Park South  Caribbean Restaurant        
48   Georgetown           Bank                        
55   Wingate              Fast Food Restaurant        
56   Rugby                Caribbean Restaurant        
57   Remsen Village       Caribbean Restaurant        
59   Paerdegat Basin      Gym                         
60   Mill Basin           Pizza Place                 
63   Weeksville           Chinese Restaurant          
66   Homecrest            Bank                        
69   Erasmus              Caribbean Restaurant        
71   Bensonhurst          Chinese Restaurant          
74   Gravesend            Italian Restaurant          
80   East Flatbush        Food & Drink Shop           
95   Cypress Hills        Latin American Restaurant   
97   Starrett City        Pizza Place                 
98   Canarsie             IT Services                 
99   Flatlands            Pharmacy                    
103  Bath Beach           Chinese Restaurant          
109  Sea Gate             Spa                         
114  City Line            Donut Shop                  
117  Prospect Park South  Caribbean Restaurant        
118  Georgetown           Bank                        
125  Wingate              Fast Food Restaurant        
126  Rugby                Caribbean Restaurant        
127  Remsen Village       Caribbean Restaurant        
129  Paerdegat Basin      Gym                         
130  Mill Basin           Pizza Place                 
133  Weeksville           Chinese Restaurant          
136  Homecrest            Bank                        
139  Erasmus              Caribbean Restaurant        

          2nd Most Common Venue            3rd Most Common Venue  \
1    Ice Cream Shop              Donut Shop                        
4    Chinese Restaurant          Bus Station                       
10   Print Shop                  Chinese Restaurant                
25   Fried Chicken Joint         Ice Cream Shop                    
27   Shopping Mall               Donut Shop                        
28   Asian Restaurant            Bus Line                          
29   Caribbean Restaurant        Fried Chicken Joint               
33   Pharmacy                    Kids Store                        
39   Construction & Landscaping  Beach                             
44   Pharmacy                    Shoe Store                        
47   Pharmacy                    Pizza Place                       
48   Breakfast Spot              Donut Shop                        
55   Deli / Bodega               Juice Bar                         
56   Bank                        Supermarket                       
57   Fast Food Restaurant        Southern / Soul Food Restaurant   
59   Moving Target               Food                              
60   Chinese Restaurant          Sushi Restaurant                  
63   Discount Store              Grocery Store                     
66   Grocery Store               Donut Shop                        
69   Yoga Studio                 Gym / Fitness Center              
71   Ice Cream Shop              Donut Shop                        
74   Chinese Restaurant          Bus Station                       
80   Print Shop                  Chinese Restaurant                
95   Fried Chicken Joint         Ice Cream Shop                    
97   Shopping Mall               Donut Shop             

#### Cluster 3

In [98]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 2, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
35   Dyker Heights  Plaza                 Bagel Shop             
105  Dyker Heights  Plaza                 Bagel Shop             

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
35   Food                  Burger Joint          Golf Course            
105  Food                  Burger Joint          Golf Course            

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
35   Field                 Falafel Restaurant    Farm                   
105  Field                 Falafel Restaurant    Farm                   

    9th Most Common Venue 10th Most Common Venue  
35   Farmers Market        Fast Food Restaurant   
105  Farmers Market        Fast Food Restaurant

#### Cluster 4

In [99]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 3, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
30   Mill Island  Pool                  Women's Store          
100  Mill Island  Pool                  Women's Store          

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
30   Filipino Restaurant   Factory               Falafel Restaurant     
100  Filipino Restaurant   Factory               Falafel Restaurant     

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
30   Farm                  Farmers Market        Fast Food Restaurant   
100  Farm                  Farmers Market        Fast Food Restaurant   

    9th Most Common Venue 10th Most Common Venue  
30   Field                 Fish & Chips Shop      
100  Field                 Fish & Chips Shop

#### Cluster 5

In [100]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 4, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood      1st Most Common Venue  \
0    Bay Ridge                  Spa                         
2    Sunset Park                Latin American Restaurant   
3    Greenpoint                 Bar                         
5    Brighton Beach             Beach                       
6    Sheepshead Bay             Dessert Shop                
8    Flatbush                   Deli / Bodega               
9    Crown Heights              Pizza Place                 
11   Kensington                 Thai Restaurant             
12   Windsor Terrace            Deli / Bodega               
13   Prospect Heights           Bar                         
14   Brownsville                Restaurant                  
15   Williamsburg               Pizza Place                 
16   Bushwick                   Bar                         
17   Bedford Stuyvesant         Deli / Bodega               
18   Brooklyn Heights           Yoga Studio                 
19   Cobble Hill                Playground                  
20   Carroll Gardens            Italian Restaurant          
21   Red Hook                   Art Gallery                 
22   Gowanus                    Bar                         
23   Fort Greene                Italian Restaurant          
24   Park Slope                 Coffee Shop                 
26   East New York              Deli / Bodega               
31   Manhattan Beach            Café                        
32   Coney Island               Pizza Place                 
34   Borough Park               Bank                        
36   Gerritsen Beach            Ice Cream Shop              
37   Marine Park                Ice Cream Shop              
38   Clinton Hill               Pizza Place                 
40   Downtown                   Burger Joint                
41   Boerum Hill                Coffee Shop                 
42   Prospect Lefferts Gardens  Bakery                      
43   Ocean Hill                 Deli / Bodega               
45   Bergen Beach               Harbor / Marina             
49   East Williamsburg          Bar                         
50   North Side                 Coffee Shop                 
51   South Side                 Bar                         
52   Ocean Parkway              Bakery                      
53   Fort Hamilton              Pizza Place                 
54   Ditmas Park                Kids Store                  
58   New Lots                   Pizza Place                 
61   Fulton Ferry               Park                        
62   Vinegar Hill               Food Truck                  
64   Broadway Junction          Diner                       
65   Dumbo                      Coffee Shop                 
67   Highland Park              Liquor Store                
68   Madison                    Deli / Bodega               
70   Bay Ridge                  Spa                         
72   Sunset Park                Latin American Restaurant   
73   Greenpoint                 Bar                         
75   Brighton Beach             Beach                       
76   Sheepshead Bay             Dessert Shop                
78   Flatbush                   Deli / Bodega               
79   Crown Heights              Pizza Place                 
81   Kensington                 Thai Restaurant             
82   Windsor Terrace            Deli / Bodega               
83   Prospect Heights           Bar                         
84   Brownsville                Restaurant                  
85   Williamsburg               Pizza Place                 
86   Bushwick                   Bar                         
87   Bedford Stuyvesant         Deli / Bodega               
88   Brooklyn Heights           Yoga Studio                 
89   Cobble Hill                Playground                  
90   Carroll Gardens            Italian Restaurant          
91   Red Hook                   Art Gallery                 
92   Gowanus                    Bar                  